### PR TITLE
Label drift fixes

### PR DESCRIPTION
**Pull request by nirhutnik**
Fixed the TrainTestLabelDrift check.
It now runs on transformers in batch.
Also, enabled custom transformers if needed.
And fixed bugs in display.

#677

### CODE REVIEW

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 1]
```diff
@@ -10,23 +10,58 @@
...
 import plotly.graph_objs as go
 
-
 __all__ = ['TrainTestLabelDrift']
 
 
+# TODO: Add label sampling when available
+
+# Functions temporarily here, will be changed when Label and Prediction classes exist:
+def get_bbox_area(label):
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:32]: @ItayGabbay These probably should be moved to your new code for format transformations.

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 2]
```diff
@@ -69,82 +104,35 @@ def run_logic(self, context: Context) -> CheckResult:
...
-                test_distribution=test_label_distribution,
-                plot_title='Class',
-                column_type='categorical',
-            )
-
-            values_dict = {'Drift score': drift_score, 'Method': method}
-            displays.append(display)
-
+        if self.alternative_label_transformers is not None:
+            label_transformers_list = self.alternative_label_transformers
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:108]: validate dict structure? maybe in init.

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 3]
```diff
@@ -69,82 +104,35 @@ def run_logic(self, context: Context) -> CheckResult:
...
-            drift_score, method, display = calc_drift_and_plot(
-                train_distribution=train_label_distribution,
-                test_distribution=test_label_distribution,
-                plot_title=title,
-                column_type='numerical'
-            )
-
-            values_dict[title] = {'Drift score': drift_score, 'Method': method}
-            displays.append(display)
+        # Drift on other custom metrics:
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:117]: other?

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 4]
```diff
@@ -49,10 +84,10 @@ class TrainTestLabelDrift(TrainTestCheck):
 
     def __init__(
             self,
-            max_num_categories: int = 10
+            alternative_label_transformers: List[Dict[str, Any]] = None
     ):
         super().__init__()
-        self.max_num_categories = max_num_categories
+        self.alternative_label_transformers = alternative_label_transformers
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:90]: Can we find an alternative to "transformers"? E.g. "properties" or "property_calculators"? Just that label_transformers may be confused with other stuff...

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 5]
```diff
@@ -10,23 +10,58 @@
...
+DEFAULT_OBJECT_DETECTION_LABEL_TRANSFORMERS = [
+    {'name': 'Bounding box area distribution', 'method': get_bbox_area, 'is_continuous': True},
+    {'name': 'Samples per class', 'method': get_samples_per_class_object_detection, 'is_continuous': False},
+    {'name': 'Number of bounding boxes per image', 'method': count_num_bboxes, 'is_continuous': False},
+]
+
+
 class TrainTestLabelDrift(TrainTestCheck):
     """
     Calculate label drift between train dataset and test dataset, using statistical measures.
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:67]: Please update docstring. Should explain the format and use of alternative_label_transformers

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 6]
```diff
@@ -69,82 +104,35 @@ def run_logic(self, context: Context) -> CheckResult:
...
 
-            values_dict[title] = {'Drift score': drift_score, 'Method': method}
+            values_dict[d['name']] = {'Drift score': drift_score, 'Method': method}
             displays.append(display)
 
-        else:
-            raise NotImplementedError('Currently not implemented')  # TODO
-
         headnote = """<span>
             The Drift score is a measure for the difference between two distributions, in this check - the test
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:137]: Also think that this text should be updated to explain what this "distribution" is.

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 7]
```diff
@@ -194,49 +182,129 @@ def condition(result: Dict) -> ConditionResult:
...
-    for batch in dataset.get_data_loader():
-        list_of_arrays = batch[1]
-        calc_res = [label_transformer(arr) for arr in list_of_arrays]
-        if len(calc_res) != 0 and isinstance(calc_res[0], list):
-            calc_res = [x[0] for x in sum(calc_res, [])]
-        counter.update(calc_res)
-    return counter
+def generate_label_histograms_by_batch(train_dataset: VisionDataset, test_dataset: VisionDataset,
+                                       label_transformers: List[Dict[str, Any]] = None,
+                                       num_bins: int = 100) -> Tuple[List[Dict[Any, float]], List[Dict[Any, float]]]:
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:187]: num_bins should be a check param.

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 8]
```diff
@@ -69,82 +104,35 @@ def run_logic(self, context: Context) -> CheckResult:
...
-                train_distribution=train_label_distribution,
-                test_distribution=test_label_distribution,
-                plot_title='Class',
-                column_type='categorical',
-            )
-
-            values_dict = {'Drift score': drift_score, 'Method': method}
-            displays.append(display)
-
+        if self.alternative_label_transformers is not None:
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:107]: use validate_shared_label() (also validate label?)

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 9]
```diff
@@ -194,49 +182,129 @@ def condition(result: Dict) -> ConditionResult:
...
+        calc_res = get_results_on_batch(batch, continuous_label_transformers[i])
+        new_hist, _ = np.histogram(calc_res, bins=num_bins, range=(bounds[i][0], bounds[i][1]))
+        hists[i] += new_hist
+    return hists
+
+
+def get_results_on_batch(batch, label_transformer):
+    """Calculate transformer result on batch of labels."""
+    list_of_arrays = batch[1]
+    calc_res = [label_transformer(arr) for arr in list_of_arrays]
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:292]: here comes the tricky part, you need to apply the label_transformer (the one from the dataset, that converts this to a common format) befor applying your label_transformer (which computes some quantity on it). That's why I think the latter should be renamed(:

File: tests/vision/checks/distribution/train_test_label_drift_test.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+# ----------------------------------------------------------------------------
+#
+"""Test functions of the VISION train test label drift."""
+from hamcrest import assert_that, has_entries, close_to, equal_to
+
+from deepchecks.vision.checks import TrainTestLabelDrift
+from tests.vision.vision_conftest import *
+
+
+def test_no_drift_classification_label(mnist_dataset_train, mnist_dataset_test):
```
Comment (noamzbr) [inline tests/vision/checks/distribution/train_test_label_drift_test.py:18]: These should obviously be expanded.

**Review by noamzbr — CHANGES_REQUESTED** [review]

File: docs/source/examples/checks/vision/distribution/train_test_label_drift.ipynb [hunk 11]
```diff
@@ -201,7 +201,7 @@
          "showlegend": false,
```
Comment (noamzbr) [inline docs/source/examples/checks/vision/distribution/train_test_label_drift.ipynb:201]: <p>Remember to update this example to our new format. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/797/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 9 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:292]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 8 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:107]: No need because of Context

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 7 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:187]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 2 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:108]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 4 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:90]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 3 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:117]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 6 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:137]: Done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 5 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:67]: Done

File: tests/vision/checks/distribution/train_test_label_drift_test.py (hunk 10 above)
Comment (nirhutnik) [reply tests/vision/checks/distribution/train_test_label_drift_test.py:18]: Will do later, as discussed

File: docs/source/examples/checks/vision/distribution/train_test_label_drift.ipynb (hunk 11 above)
Comment (nirhutnik) [reply docs/source/examples/checks/vision/distribution/train_test_label_drift.ipynb:201]: Will do later, as discussed

**Review by noamzbr — APPROVED** [review]
